From 7e64bbaa39494600bf9cd426d043f1f908c5a36c Mon Sep 17 00:00:00 2001
From: WANG Xuerui <git@xen0n.name>
Date: Sat, 8 May 2021 16:44:51 +0800
Subject: [PATCH 08/10] LoongArch: allow configuring GCC without
 --with-arch=loongarch arg

Also add helpful error messages.

    gcc/
    * config.gcc: Modify LoongArch config.
---
 gcc/config.gcc | 40 ++++++++++++++++++++++++++++++++--------
 1 file changed, 32 insertions(+), 8 deletions(-)

diff --git a/gcc/config.gcc b/gcc/config.gcc
index fd32208cec0..e8f8c3cf976 100644
--- a/gcc/config.gcc
+++ b/gcc/config.gcc
@@ -451,7 +451,7 @@ loongarch*-*-*)
 	extra_headers="larchintrin.h"
 	extra_objs="loongarch-c.o loongarch-builtins.o"
 	extra_options="${extra_options} g.opt fused-madd.opt loongarch/loongarch-tables.opt"
-  ;;
+	;;
 nds32*)
 	cpu_type=nds32
 	extra_headers="nds32_intrinsic.h nds32_isr.h nds32_init.inc"
@@ -4935,6 +4935,36 @@ case "${target}" in
 	loongarch*-*-*)
 		supported_defaults="abi arch float fpu tune fix-loongson3-llsc"
 
+		case ${with_arch} in
+		loongarch | loongarch64)
+			# OK
+			default_loongarch_arch=loongarch64
+			;;
+		"")
+			# fallback
+			default_loongarch_arch=loongarch64
+			;;
+		*)
+			echo "Unknown arch given in --with-arch=$with_arch, available choices are: loongarch64" 1>&2
+			exit 1
+			;;
+		esac
+
+		case ${with_abi} in
+		lp64 | lp32)
+			# OK
+			default_loongarch_abi=$with_abi
+			;;
+		"")
+			# fallback
+			default_loongarch_abi=lp64
+			;;
+		*)
+			echo "Unknown ABI given in --with-abi=$with_abi, available choices are: lp32 lp64" 1>&2
+			exit 1
+			;;
+		esac
+
 		case ${with_float} in
 		"" | soft | hard)
 			# OK
@@ -5419,14 +5449,8 @@ case ${target} in
 		;;
 
 	loongarch*-*-*)
-		if test x$with_arch != x; then
-			default_loongarch_arch=$with_arch
-		fi
-		if test x$with_abi != x; then
-			default_loongarch_abi=$with_abi
-		fi
 		case ${default_loongarch_arch} in
-		    loongarch | loongarch64) tm_defines="$tm_defines LARCH_ISA_DEFAULT=0" ;;
+		    loongarch64) tm_defines="$tm_defines LARCH_ISA_DEFAULT=0" ;;
 		esac
 		case ${default_loongarch_abi} in
 		    lp64)   tm_defines="$tm_defines LARCH_ABI_DEFAULT=ABILP64" ;;
-- 
2.30.1

